src/infra/flow-system/inmem/src/flow_trigger/inmem_flow_trigger_event_store.rs

```diff
@@ -0,0 +1,107 @@
+
+pub struct InMemoryFlowTriggerEventStore {
+    inner: InMemoryEventStore<FlowTriggerState, State>,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+struct State {
+    events: Vec<FlowTriggerEvent>,
+    dataset_ids: Vec<DatasetID>,
+}
+
+impl EventStoreState<FlowTriggerState> for State {
+    fn events_count(&self) -> usize {
+        self.events.len()
+    }
+
+    fn get_events(&self) -> &[FlowTriggerEvent] {
+        &self.events
+    }
+
+    fn add_event(&mut self, event: FlowTriggerEvent) {
+        self.events.push(event);
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[component(pub)]
+#[interface(dyn FlowTriggerEventStore)]
+#[scope(Singleton)]
+impl InMemoryFlowTriggerEventStore {
+    pub fn new() -> Self {
+        Self {
+            inner: InMemoryEventStore::new(),
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl EventStore<FlowTriggerState> for InMemoryFlowTriggerEventStore {
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn len(&self) -> Result<usize, InternalError> {
+        self.inner.len().await
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, fields(?query, ?opts))]
+    fn get_events(&self, query: &FlowKey, opts: GetEventsOpts) -> EventStream<FlowTriggerEvent> {
+        self.inner.get_events(query, opts)
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, fields(?query, num_events = events.len()))]
+    async fn save_events(
+        &self,
+        query: &FlowKey,
+        maybe_prev_stored_event_id: Option<EventID>,
+        events: Vec<FlowTriggerEvent>,
+    ) -> Result<EventID, SaveEventsError> {
+        if events.is_empty() {
+            return Err(SaveEventsError::NothingToSave);
+        }
+
+        if let FlowKey::Dataset(flow_key) = query {
+            let state = self.inner.as_state();
+            let mut g = state.lock().unwrap();
+            g.dataset_ids.push(flow_key.dataset_id.clone());
+        }
+
+        self.inner
+            .save_events(query, maybe_prev_stored_event_id, events)
+            .await
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl FlowTriggerEventStore for InMemoryFlowTriggerEventStore {
+    #[tracing::instrument(level = "debug", skip_all)]
+    fn list_all_dataset_ids(&self) -> FailableDatasetIDStream {
+        use futures::StreamExt;
+
+        let dataset_ids = self.inner.as_state().lock().unwrap().dataset_ids.clone();
+
+        // TODO: re-consider performance impact
+        Box::pin(tokio_stream::iter(dataset_ids).map(Ok))
```

Comment:
I can predict the impact. Plz implement pagination from the start.
The same applies to Postgres and SQLite - I think pagination should come from service level.